### PR TITLE
cpu/saml21: Make Low-Power SRAM available to programs

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

SAML21 contains up to 8 kiB low-power SRAM that can be retained during deep sleep modes.

This adds a section for this memory in the linker script, so structures can be placed in it by giving them the `__attribute__((section(".backup")))`

e.g.:

    static uint8_t persistent_buffer[64] __attribute__((section(".backup")));

Compile-tested only, run-tested on same54-xpro which also has 8 kiB of Backup RAM.

### Testing procedure

````
static uint8_t persistent_buffer[64] __attribute__((section(".backup")));

int main(void) {
    /* simple check if the memory is still the same after reboot */
    if (persistent_buffer[0] != 0x23) {
            persistent_buffer[0] = 0x23;
            puts("fresh memory");
        } else {
            puts("retained memory");
    }

    /* let the RTC trigger a reset after 5s */
    struct tm time;
    rtc_get_time(&time);
    time.tm_sec += 5;
    mktime(&time);
    rtc_set_alarm(&time, NULL, NULL);

    /* this will trigger a deep sleep */
    pm_unblock(0);
    xtimer_usleep(10000);

    return 0;
}
````

### Issues/PRs references

I'm doing the same in #11305 as SAMD5x has a very similar feature (Backup Ram instead of Low-Power SRAM but the idea is the same).

### CODE REVIEW

File: cpu/saml21/Makefile.include [hunk 1]
```diff
@@ -3,5 +3,17 @@ export CPU_FAM  = saml21
 
 USEMODULE += pm_layered
 
+ifneq (,$(filter saml21j18b saml21j18a,$(CPU_MODEL)))
```
Comment (dylad) [inline cpu/saml21/Makefile.include]: Can you also add SAMR30 here please ? We might find a better way for this later to avoid adding new MCU.

File: cpu/saml21/ldscripts/saml21.ld [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @addtogroup      cpu_saml21
+ * @{
+ *
+ * @file
+ * @brief           Memory definitions for Atmel SAML21
```
Comment (dylad) [inline cpu/saml21/ldscripts/saml21.ld]:
```suggestion
 * @brief           Memory definitions for Microchip SAML21
```

File: cpu/saml21/Makefile.include (hunk 1 above)
Comment (benpicco) [reply cpu/saml21/Makefile.include]: Ok.

Comment (aabadie) [reply cpu/saml21/Makefile.include]: Maybe
```mk
ifneq (,$(filter saml21,$(CPU_FAM)))
```
?

File: cpu/saml21/ldscripts/saml21.ld (hunk 2 above)
Comment (benpicco) [reply cpu/saml21/ldscripts/saml21.ld]: Well even my shiny new same54 still has Atmel printed on it, but names are sound and smoke 😉

File: cpu/saml21/Makefile.include (hunk 1 above)
Comment (benpicco) [reply cpu/saml21/Makefile.include]: They don't all have the same amount.
SAML21x18 & SAML21x17 have 8k, SAML21x16 has 4k and SAML21E15 comes with 2k LP-SRAM.

File: cpu/saml21/ldscripts/saml21.ld (hunk 2 above)
Comment (dylad) [reply cpu/saml21/ldscripts/saml21.ld]: True ! But my contacts at Microchip kindly ask me to rename everything 😉

File: cpu/saml21/ldscripts/saml21.ld [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ccmram_length = DEFINED( _lpsram_len ) ? _lpsram_len : 0x0 ;
+
+MEMORY
+{
+    lp-sram  : ORIGIN = 0x30000000, LENGTH = _lpsram_len
+}
+
+SECTIONS
+{
+    .bkup : {
```
Comment (jcarrano) [inline cpu/saml21/ldscripts/saml21.ld]: For clarity, I'd rather have the output section named ".backup" too.

Comment (jcarrano) [inline cpu/saml21/ldscripts/saml21.ld]: This is missing (NOLOAD). Without it, I get a __769M binary__!!!!

**Comment by jcarrano** [comment]
@benpicco I PR'd against your branch.

File: cpu/cortexm_common/ldscripts/cortexm.ld [hunk 4]
```diff
@@ -23,12 +23,16 @@
...
+_backup_ram_start_addr = DEFINED( _backup_ram_start_addr ) ? _backup_ram_start_addr : 0x0 ;
+_backup_ram_len = DEFINED( _backup_ram_len ) ? _backup_ram_len : 0x0 ;
+
 ASSERT((_fw_rom_length <= _rom_length - _rom_offset), "Specified firmware size does not fit in ROM");
 
 MEMORY
 {
     rom (rx)    : ORIGIN = _rom_start_addr + _rom_offset, LENGTH = _fw_rom_length
     ram (w!rx)  : ORIGIN = _ram_start_addr,               LENGTH = _ram_length
+    bkup-ram    : ORIGIN = _backup_ram_start_addr,        LENGTH = _backup_ram_len
```
Comment (jcarrano) [inline cpu/cortexm_common/ldscripts/cortexm.ld]:
```suggestion
    bkup-ram (w!rx) : ORIGIN = _backup_ram_start_addr,    LENGTH = _backup_ram_len
```

Also, I think this new approach is better than having a separate ldscript. With a default length of 0, any attemt to use ".backup" where it is not supported will cause a link failure.

Comment (benpicco) [reply cpu/cortexm_common/ldscripts/cortexm.ld]: fixed.

Comment (jcarrano) [reply cpu/cortexm_common/ldscripts/cortexm.ld]: The `(w!rx)` is of course just a nitpick.

File: examples/hello-world/main.c [hunk 5]
```diff
@@ -20,13 +20,29 @@
...
-    printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
-    printf("This board features a(n) %s MCU.\n", RIOT_MCU);
+    printf("counter value, %d\n", counter);
+    printf("counter_zero value, %d\n", counter_zero);
+
+    while (1) {
+        printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
+        printf("This board features a(n) %s MCU.\n", RIOT_MCU);
+        printf("counter value, %d\n", counter++);
+        printf("counter_zero value, %d\n", counter_zero++);
```
Comment (dylad) [inline examples/hello-world/main.c]: Isn't better to create a dedicated test instead of temporarily hacked this one ?
This may be useful to newcomers that need this feature.

Comment (jcarrano) [reply examples/hello-world/main.c]: Yes, of course (in fact it was mentioned in a short discussion on ben's branch). That, along with a documentation page.

Comment (benpicco) [reply examples/hello-world/main.c]: I've added a dedicated (albeit pretty simple) test.
How should I go about the documentation?

**Comment by jcarrano** [comment]
Looks good. I'll test it as soon as I can get a saml21.

On another note - and this is a topic for another PR- I think that this approach of making the liker aware of backup memory regions is useful even when that memory is not transparently usable.

I'm referring to chips like the TM4C123. That part has 64 bytes of battery backed memory _that can't be read like normal memory_ because it lives on a different clock domain. Even then it is still convenient to declare that region in the linker, as it allows different components of an application to claim a portion to themselves without worrying about overlap and without having resort to centrally defined mapping (like in a header).

**Comment by benpicco** [comment]
I reworded the commit message.

**Comment by benpicco** [comment]
@jcarrano ping

**Comment by jcarrano** [comment]
@dylad, @aabadie : any of you has the device to test this. Code-wise this is fine for me, but I'd like if someone can test it.

Also, I'm dismissing my review because the issues have been addressed. I'll appove as soon as there is confirmation that it works.

**Comment by dylad** [comment]
bad news, test doesn't work.
```
019-06-12 20:56:47,576 - INFO # 
2019-06-12 20:56:47,577 - INFO # Backup RAM test
2019-06-12 20:56:47,578 - INFO # 
2019-06-12 20:56:47,584 - INFO # This test will increment the counter by 1, then enter deep sleep for 1s
2019-06-12 20:56:47,585 - INFO # counter: 1
```
And no more output, impossible to reset or flash, I can only unplug/replug and test get stuck again. I'll try to find some times to check what happen.

File: tests/periph_backup_ram/main.c [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+ * @}
+ */
+
+#include <stdio.h>
+
+#include "periph/rtc.h"
+#include "pm_layered.h"
+
+#ifndef SEEP_SEC
+#define SEEP_SEC 1
```
Comment (dylad) [inline tests/periph_backup_ram/main.c]: SLEEP_SEC ?
Here and everywhere

**Comment by benpicco** [comment]
That is on saml21 I suppose?

I've tried it today too on same54 only to find out that the RTC would not work - I then flashed `tests/periph_rtt` and `tests/periph_rtc` and suddenly it would work again. Non-deterministic bugs are the worst, so I began [cleaning up RTT & RTC](https://github.com/benpicco/RIOT/commits/sam0_32k-gclk) again. Do you get consistent behavior with that branch? I'll have to do more tests tomorrow.

I skimmed the data sheet and indeed it looks like LP SRAM is for DMA during Standby? Well I guess I'll then drop saml21 from this patch :confused: 

Might still be useful to have a section for that memory though.

**Comment by dylad** [comment]
> That is on saml21 I suppose?

Yes, I only tested on this board for now.
The test hangs in the same way everytime for me on SAML21.
For SAML21, seems like you can only use EXTI and RTC peripherals in backup mode.

> Might still be useful to have a section for that memory though.

Indeed, you can do a lot of thing with a few KB :)

**Comment by benpicco** [comment]
Is `tests/periph_rtc` working for you?

Also, I think I should add an `xtimer_sleep(5);` in there - when the MCU is in backup mode, I also have trouble flashing it with J-Link, there should be a grace period of no-sleep to allow for easy re-flashing.

**Comment by dylad** [comment]
> Is tests/periph_rtc working for you?

Yes on both SAME54-XPRO and SAML21-XPRO (only tested your branch but must be good on master too)

> I also have trouble flashing it with J-Link

You're not using EDBG ? If you're stuck a quick `edbg -t atmel_cm0p -e` will save you.

**Comment by benpicco** [comment]
But EDBG doesn't work with SWD, does it? With the dev-boards I have no problem flashing them when they are in deep sleep, but with the 'real' board I only have the SWD interface where I use OpenOCD with J-Link (I think ST-Link should work too).

**Comment by dylad** [comment]
@benpicco 
Indeed, EDBG will not work with SWD but you can try to use JLinkExe directly instead of openOCD. I never had any issue with JLinkExe and low power mode on SAM0.

**Comment by dylad** [comment]
@benpicco is #11689 needed for this PR ? I'd like to move forward on this.
Even if this test cannot work for SAML21, LP SRAM is usefull (I really like to introduce DMA transfer for SPI transaction).
Did you solve your current issues ?

**Comment by dylad** [comment]
Could you rebased btw ? I'll give a try on SAME54.

**Comment by dylad** [comment]
forget my previous message, I forget to pull lastest changes on this computer...

**Comment by benpicco** [comment]
#11689 should only be needed if the RTC was buggy before.

I'm unsure what do to about saml21. Should I just drop 0507318e21fd64569c86a695eea3112027cf6d26 or should I add a comment to the Makefile that the Low Power SRAM is **not** retained in Backup mode? That'd make me re-think the interface specification if already half the platforms that initially support it don't implement it properly.

**Comment by benpicco** [comment]
I moved it to `periph_cpu.h`.

File: cpu/cortexm_common/vectors_cortexm.c [hunk 7]
```diff
@@ -78,7 +88,7 @@ __attribute__((weak)) void post_startup (void)
 void reset_handler_default(void)
 {
     uint32_t *dst;
-    uint32_t *src = &_etext;
+    const uint32_t *src = &_etext;
```
[the remark this answers is not in the document]
Comment (benpicco) [reply cpu/cortexm_common/vectors_cortexm.c]: I can remove it.
But `.text` *should* be `const`.

**Comment by benpicco** [comment]
I think a rebase is in order - can I squash while I'm at it?

**Comment by dylad** [comment]
@benpicco feel free to squash

File: tests/periph_backup_ram/main.c [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        puts("WARNING: non-backup memory retained - did we really enter deep sleep?");
+    }
+
+    printf("counter: %d\n", ++counter);
+    counter_noinit = counter;
+
+    /* Some tools have trouble flashing MCUs in deep sleep.
+     * Wait a bit to make re-flashing / debugging easier.
+     */
+    xtimer_sleep(3);
```
Comment (dylad) [inline tests/periph_backup_ram/main.c]: This traditional delay is troublesome because you're suppose to see the counter increments by one every second but currently this is not the case.

File: tests/periph_backup_ram/main.c [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    static int counter __attribute__((section(".backup.bss")));
+
+    if (counter == 0) {
+        puts("\nBackup RAM test\n");
+        printf("This test will increment the counter by 1, "
+               "then enter deep sleep for %ds\n", SLEEP_SEC);
+    } else if (counter_noinit == counter) {
+        puts("WARNING: non-backup memory retained - did we really enter deep sleep?");
+    }
+
```
Comment (dylad) [inline tests/periph_backup_ram/main.c]: My personal taste would have place `#ifndef CPU_BACKUP_RAM_NOT_RETAINED` here so SAML21 platform doesn't display the counter before saying "aborting test". You may ignore this comment if you're fine this way.

File: tests/periph_backup_ram/main.c (hunk 8 above)
Comment (benpicco) [reply tests/periph_backup_ram/main.c]: > you're suppose to see the counter increments by one every second 

Says who? :wink: 
I've added some output to clarify the delay periods.

It's really annoying when you have OpenOCD and your MCU enters deep sleep with no delay - there is no way to re-flash it. (If you're lucky you have JLink or edbg. But I'd rather save the unsuspecting Raspberry Pi GPIO users the cursing.)

File: tests/periph_backup_ram/main.c (hunk 9 above)
Comment (benpicco) [reply tests/periph_backup_ram/main.c]: I think we should at least once access that memory to make sure it doesn't crash & burn when doing so.

File: tests/periph_backup_ram/main.c (hunk 8 above)
Comment (dylad) [reply tests/periph_backup_ram/main.c]: I don't mind keeping this delay but at least update the printf so users wont think this test is too slow ;)_

File: tests/periph_backup_ram/main.c [hunk 10]
```diff
@@ -51,15 +58,14 @@ int main(void)
...
-    xtimer_sleep(3);
+    xtimer_sleep(DELAY_SEC);
 
 #ifndef CPU_BACKUP_RAM_NOT_RETAINED
 
     /* schedule RTC wake in SLEEP_SEC s */
     struct tm time;
     rtc_get_time(&time);
     time.tm_sec += SLEEP_SEC;
-    mktime(&time);
```
Comment (dylad) [inline tests/periph_backup_ram/main.c]: Why did you remove this line ??

Comment (benpicco) [reply tests/periph_backup_ram/main.c]: Because #11413 got merged.

**Comment by dylad** [comment]
please squash !

**Comment by benpicco** [comment]
squashed.

**Comment by benpicco** [comment]
Thank you for the review and testing!
